Jetty 9.4.x 4331 async close complete3

gregw opened the pull request:
This replaces #4398 as a fix for #4331 

This PR breaks out the HttpOutput state into api state and close state.

gregw committed:
Issue #4376 Async Content Complete

Added test harness to reproduce unready completing write.
Fixed test by not closing output prior to becoming READY

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Issue #4331 Async Write Complete

Test harness to reproduce unready when closing/completing.

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Issue #4331 Async Write Complete

test both PENDING and UNREADY

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Issue #4331 Async Write Complete

test cleanups

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Issue #4331 Async Close Complete

Cleanups of write

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Merge remote-tracking branch 'origin/jetty-9.4.x' into jetty-9.4.x-4331-asyncCloseComplete2

gregw committed:
WIP

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Issue #4331 Close Complete

Work in progress

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Issue #4331 Close Complete

Added async close to HttpWriter and ResponseWriter
Always use async close, with blocker if necessary.

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Issue #4331 Close Complete

Working async close complete!

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Issue #4331 Close Complete

invert test as we can now call complete when not ready!

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Issue #4331 Close Complete

fixed transition to ERROR state

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Issue #4331 Close Complete

async close after onError

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Issue #4331 Close Complete

minor cleanups

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Issue #4331 Close Complete

Fix for proxy tests

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Issue #4331 Close Complete

Fix write loop to handle clear of p=0,l=0 rather than p=l

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Issue #4331 Close Complete

Removed old close on all content mechanism
Cleanups and some more TODOs

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Issue #4331 Close Complete

a reworking of HttpOutput to separate out API state.

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Issue #4331 Close Complete

Soft close for Dispatcher
release buffer in onWriteComplete

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Issue #4331 Close Complete

Set _onError in onWriteComplete
NOOP callback instead of null

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Issue #4331 Close Complete

failure closes HttpOutput

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Issue #4331 Close Complete

Moved closedCallback handling to onWriteComplete

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Issue #4331 Close Complete

Additional test of complete during blocking write.

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Issue #4331 Close Complete

reimplemented blocking close to sometimes be async

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Issue #4331 Close Complete

ascii "art"

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Issue #4331 Close Complete

Code cleanup.  Use a CLOSE state rather than non null closedCallback to be clearer that it is a state.
Renamed close(Callback) to complete(Callback)
Renamed and simplified closed() to completed()

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Issue #4331 Close Complete

Do not dispatch
Better ascii art
improved close impl to be similar to complete

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Issue #4331 Close Complete

More test cases

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Issue #4331 Close Complete

retain execute behaviour in 9.4. review in 10.

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw commented:
@lachlan-roberts @sbordet This is ready for proper review.  Would be good to get a green tick early Monday so we can start a release cycle.

lachlan-roberts commented on jetty-server/src/main/java/org/eclipse/jetty/server/HttpOutput.java:
```diff
@@ -63,28 +60,82 @@
+      w = write
+      owc = onWriteComplete(false,null)
+      owcL = onWriteComplete(true,null)
+      isf = isReady()==false
```
This is different to what is on the diagram. This should be irf and irt?

gregw replied:
yes - fixed

gregw committed:
Improved javadoc and ascii art

gregw committed:
Improved CLOSING

Switch to CLOSING state as soon as last write is done, even if several non last channelWrites will be done.   This allows a subsequent call to close to know that nothing needs to be written and can avoid some EOF exceptions. Now onWriteComplete acts only on the passed in last parameter.

Added test for sendContent

gregw committed:
WIP

Aggregate within lock
pipeline test debug

gregw committed:
Avoid creating ignored exception when Idle or Failed.

gregw committed:
Try a parse without fill to avoid unconsumed input debug

gregw committed:
fixed pipeline size

gregw commented:
@sbordet @lachlan-roberts I'm am getting the failure in the attached log very occassionally... only one one laptop... only once in about 1000 runs... but sometimes 10 times in a row!!!
I have no idea what is going on at the moment, so would definitely appreciate a debugging duck at some stage.   But can either of you reproduce on your machines on this branch?
[DelayedServerTest_testPipeline.txt](https://github.com/eclipse/jetty.project/files/3966298/DelayedServerTest_testPipeline.txt)

gregw committed:
release buffer before callback

gregw commented:
@sbordet @lachlan-roberts I think I found the problem... I was calling the closedCallback and then releasing the buffer.  I've inverted that order and now it does not fail.

gregw committed:
turn off debug

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Issue #4331 Close Complete

Better javadoc
refactored onWriteComplete logic to be simpler
fixed bug with flush of last written byte

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Issue #4331 Close Complete

Completely reworked test harness for better coverage.

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw commented:
92% code coverage of HttpOutput with current tests

gregw commented:
@sbordet review please

gregw committed:
Merge remote-tracking branch 'origin/jetty-9.4.x' into jetty-9.4.x-4331-asyncCloseComplete3

gregw committed:
Issue #4331 Close Complete

Reworked order of ifs to match logic above in onWriteComplete

Signed-off-by: Greg Wilkins <gregw@webtide.com>
